InnerEye/ML/metrics.py

```diff
@@ -32,6 +36,140 @@
+        :param hue: The hue to restrict the values used for computation, otherwise all values will be used.
+        :returns: Tuple(optimal_threshold, false positive rate, false negative rate, accuracy)
+        """
+        if torch.unique(targets).numel() == 1:
+            return torch.tensor(np.nan), torch.tensor(np.nan), torch.tensor(np.nan), torch.tensor(np.nan)
+        fpr, tpr, thresholds = roc(preds, targets)
+        optimal_idx = torch.argmax(tpr - fpr)
+        optimal_threshold = thresholds[optimal_idx]
+        acc = accuracy(preds > optimal_threshold, targets)
+        false_negative_optimal = 1 - tpr[optimal_idx]
+        false_positive_optimal = fpr[optimal_idx]
+        return optimal_threshold, false_positive_optimal, false_negative_optimal, acc
+
+
+class AccuracyAtOptimalThreshold(ScalarMetricsBase):
+    def __init__(self, dist_sync_on_step=False):
+        super().__init__(dist_sync_on_step=dist_sync_on_step,
+                         name=MetricType.ACCURACY_AT_OPTIMAL_THRESHOLD.value)
+
+    def compute(self):
+        preds = torch.cat(self.preds)
+        target = torch.cat(self.targets)
+        return self._get_metrics_at_optimal_cutoff(preds=preds, targets=target)[3]
+
+class OptimalThreshold(ScalarMetricsBase):
+    def __init__(self, dist_sync_on_step=False):
+        super().__init__(dist_sync_on_step=dist_sync_on_step, name=MetricType.OPTIMAL_THRESHOLD.value)
+
+    def compute(self):
+        preds = torch.cat(self.preds)
+        target = torch.cat(self.targets)
+        return self._get_metrics_at_optimal_cutoff(preds=preds, targets=target)[0]
+
+
+class FalsePositiveRateOptimalThreshold(ScalarMetricsBase):
+    def __init__(self, dist_sync_on_step=False):
+        super().__init__(dist_sync_on_step=dist_sync_on_step,
+                         name=MetricType.FALSE_POSITIVE_RATE_AT_OPTIMAL_THRESHOLD.value)
+
+    def compute(self):
+        preds = torch.cat(self.preds)
+        target = torch.cat(self.targets)
+        return self._get_metrics_at_optimal_cutoff(preds=preds, targets=target)[1]
+
+
+class FalseNegativeRateOptimalThreshold(ScalarMetricsBase):
+    def __init__(self, dist_sync_on_step=False):
+        super().__init__(dist_sync_on_step=dist_sync_on_step,
+                         name=MetricType.FALSE_NEGATIVE_RATE_AT_OPTIMAL_THRESHOLD.value)
+
+    def compute(self):
+        preds = torch.cat(self.preds)
+        target = torch.cat(self.targets)
+        return self._get_metrics_at_optimal_cutoff(preds=preds, targets=target)[2]
+
+
+class AreaUnderRocCurve(ScalarMetricsBase):
+    def __init__(self, dist_sync_on_step=False):
+        super().__init__(dist_sync_on_step=dist_sync_on_step,
+                         name=MetricType.AREA_UNDER_ROC_CURVE.value)
+
+    def compute(self):
+        preds = torch.cat(self.preds)
+        targets = torch.cat(self.targets)
+        if torch.unique(targets).numel() == 1:
+            return torch.tensor(np.nan)
```

Comment:
is there a specific reason for using nan here? Did we do that in the old code too? If not, how about returning 0 if prediction is not correct, and 1 otherwise

Reply:
So sklearn pr_curve (which we used previously) returns a warning saying "hey what you are computing does not make sense you have only one class". But the PL version raises an error when this occurs. Since the metrics does not make sense anyway I thought it may be safer to return nan than a value that does not really have a meaning

Reply:
I see, yes, here we'd only see NaN if the whole, say, validation set is a single element.